sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -76,25 +79,94 @@ case class CreateTableAsSelectExec(
 
       throw new TableAlreadyExistsException(ident)
     }
+    catalog match {
+      case txnCatalog: TransactionalTableCatalog =>
```

Comment:
Because so little is shared between the two implementations, I think I would probably separate them into different exec nodes. An added benefit of that is that the physical plan would tell users whether Spark is going to use an atomic operation. That way I could check `EXPLAIN` and run if it is atomic or do more testing if it is not.